Add docs for Teleport Connect (#14458)
gravitational/teleport: docs/pages/use-teleport/teleport-connect.mdx

---
title: Using Teleport Connect
description: Using Teleport Connect
---

Teleport Connect provides easy and secure access to SSH servers and databases, with support for

> other resources such as Kubernetes clusters and application access coming in the future.

Reviewer: ```suggestion
other resources such as Kubernetes clusters and applications coming in the future.
```

![resources tab in Teleport Connect](../../img/use-teleport/connect-cluster.png)

> ## Installation & Upgrade

Reviewer: This and other headings should be in sentence case, with only the first letter and any proper nouns capititalized

Author: > This and other headings should be in sentence case, with only the first letter and any proper nouns capititalized

Thanks! I wasn't sure and I usually saw most of my US-based peers use title case for those.

Head over to the [Downloads](https://goteleport.com/download/) page to download the most recent
version of Teleport Connect. As of now, Teleport Connect supports macOS only. Double-click the

> downloaded .dmg file and drag the Teleport Connect icon to the Applications folder.

Reviewer: ```suggestion
downloaded `.dmg` file and drag the Teleport Connect icon to the Applications folder.
```

To upgrade Teleport Connect to a newer version, drag the new version to the Applications folder.

## Interface Overview

![interface overview of Teleport Connect](../../img/use-teleport/connect-ui-overview.png)

The top bar of Teleport Connect consists of:

> - The profile selector (the top right) allowing you to switch between profiles on different Teleport

Reviewer: ```suggestion
- The **profile selector** (the top right), which allows you to switch between profiles on different Teleport
```

  clusters as well as log in or out of the clusters.

> - The connection list (the top left) showing recent connections, allowing you to seamlessly switch

Reviewer: ```suggestion
- The **connection list** (the top left) showing recent connections, allowing you to seamlessly switch
```

  between them.

> - The command bar (in the middle) which launches any command you input in a new terminal tab.

Reviewer: ```suggestion
- The **command bar** (in the middle), which launches any command you input in a new terminal tab.
```

> - The cluster selector (to the left of the command bar) shows up only if there are leaf clusters

Reviewer: ```suggestion
- The **cluster selector** (to the left of the command bar), which shows up only if you have set up Trusted Clusters and there are leaf clusters
```

  connected to the root cluster. It lets you browse leaf cluster resources. It also changes
  which cluster the commands in the command bar are going to target.

> ## Connecting to an SSH Server

Reviewer: It would be great to have more UI screenshots in the sections below this point:
- Choosing an SSH server to connect to
- Choosing a database to connect to
- A relevant screenshot for the "GUI client" section
- A relevant screenshot for the "CLI Client" section
- Choosing a cluster to connect to (in "Connecting to multiple clusters")

If adding these screenshots won't be possible in the time we have, we can open a GitHub issue to do this later.

Author: > If adding these screenshots won't be possible in the time we have, we can open a GitHub issue to do this later.

Yeah, unfortunately I definitely won't have time to add those screenshots right now.

In general, we wanted keep the screenshots to minimum as we plan to substantially change the UI in the future. But that probably won't happen until another month or two if not more.

Reviewer: Makes sense! We can hold off on the screenshots for this release

connect to a server.

## Connecting to a Database

1. Open a tab with cluster resources by clicking on the plus symbol at the end of the tab bar. You
   can also press `Cmd + T` to achieve the same result.
2. Select the Databases section.
3. Look for the database server you wish to connect to and click the Connect button to the right.
4. Select or enter the database user you with to use and press `Enter`.
5. A new tab will open with a new connection established between your device and the database
   server.

This connection will remain active until you click the Close Connection button or close Teleport
Connect. The port number will persist between app restarts – you can set up your favorite client
without worrying about the port suddenly changing.

### With a GUI Client

To connect with a GUI client, follow the instructions in the database connection tab under the
Connect with GUI section.

### With a CLI Client

The database connection tab shows the command that can be used to connect to the database. You can
modify the database name of the connection and then click the Run button to open a new
terminal tab with that command executed.

## Connecting to Multiple Clusters

> After you add your first cluster, you can keep adding more by opening the profile selector in the

Reviewer: What does "After you add your first cluster" mean? Add your first leaf cluster to your root cluster? Or does this relate to the UI?

Author: > What does "After you add your first cluster" mean? Add your first leaf cluster to your root cluster? Or does this relate to the UI?

What it really refers to is logging in to a cluster. Tbh our UI right now says "Add another cluster" but what it really should say is just "Log in to another cluster". I'll update this part along with the UI.

Author: Actually, I'll change just the docs for now.

top right and clicking at the *+ Add another cluster* button.

At the moment Teleport Connect supports only one user per cluster. To log in as a different user,
log out of the cluster first.

## Restarting and Reconnecting

Before closing, Teleport Connect will remember the tabs that you had open at the end of the session.
Next time you open the app, Connect will ask you if you want to reopen those tabs. If you agree,
Connect will restore connections to all resources that were active before you closed the app.

> At the moment, terminal tabs are restored only as far as the working directory is concerned. Teleport

Reviewer: > At the moment, terminal tabs are restored only as far as the working directory is concerned.

Is there another way to phrase "as far as ... is concerned" here?

Connect doesn't attempt to re-execute commands that were in progress when the app was closed.

## Troubleshooting

> In an attempt to resolve issues with a cluster you can try logging out of it, closing the app and

Reviewer: What kinds of issues would this be?

Author: I was thinking of two types of issues:

1. A bug which results in the UI getting into a state which renders the app unusable, except for the top bar. For example, https://github.com/gravitational/webapps.e/issues/277.
2. An external change to the configuration of the cluster which renders the current profile unusable with the cluster.
   * It's hard for me to think of an example off the top of my head. But I've ran into issues in development where changing something about my Teleport cluster would render the profile stored by tsh & Connect invalid with no possibility of logging in again.
   * Usually that would be a result of doing something related to CA certs maybe, something that would be rather unlikely to happen in a live cluster but still possible.

I can try to do a general description of those issues here.

logging in again.

To force the app to log you out of all clusters, close the app and remove the `~/Library/Application
Support/Teleport Connect/tsh` folder. Removing the file under `~/Library/Application
Support/Teleport Connect/app_state.json` will clear all remembered tabs and connections.

### Submitting an Issue

To submit an issue, click the Submit Feedback button at the bottom right (the speech bubble symbol)
and follow the *Submit a Bug* link.

Be sure to attach logs which can be found under `~/Library/Application Support/Teleport
Connect/logs`. The version of the app can be found in the app menu under the About Teleport Connect
menu item.

### Updating local shell environment

> Teleport Connect updates the environment variables on app restart and not when starting a new shell

Reviewer: > the environment variables

Which environment variables does this refer to?

Author: It refers to whatever env variables are set as a result of running your shell startup files.

This should really say just "local shell environment" rather than "environment variables" specifically.

---
title: Using Teleport Connect
description: Using Teleport Connect
---

Teleport Connect provides easy and secure access to SSH servers and databases, with support for
other resources such as Kubernetes clusters and application access coming in the future.

![resources tab in Teleport Connect](../../img/use-teleport/connect-cluster.png)

## Installation & Upgrade

Head over to the [Downloads](https://goteleport.com/download/) page to download the most recent
version of Teleport Connect. As of now, Teleport Connect supports macOS only. Double-click the
downloaded .dmg file and drag the Teleport Connect icon to the Applications folder.

To upgrade Teleport Connect to a newer version, download the .dmg file with that version and follow

> the installation steps.

Reviewer 2: Best to just say drag to the Applications folder

> ## Interface Overview

Reviewer: I think "User Interface" would make sense for this heading

![interface overview of Teleport Connect](../../img/use-teleport/connect-ui-overview.png)

The top bar of Teleport Connect consists of:

- The profile selector (the top right) allowing you to switch between profiles on different Teleport
  clusters as well as log in or out of the clusters.
- The connection list (the top left) showing recent connections, allowing you to seamlessly switch
  between them.
- The command bar (in the middle) which launches any command you input in a new terminal tab.
- The cluster selector (to the left of the command bar) shows up only if there are leaf clusters
  connected to the root cluster. It lets you browse leaf cluster resources. It also changes
  which cluster the commands in the command bar are going to target.

## Connecting to an SSH Server

> 1. Open a tab with cluster resources by clicking on the plus symbol at the end of the tab bar. You

Reviewer 3: should read at the `right` end I think. Even tho there isn't one on the left, they might search there first unknowingly. Or something like 'the right most button' Idk how to word. But its on the right!

   can also press `Cmd + T` to achieve the same result.
2. Look for the SSH server you want to connect to and click the Connect button to the right.

> 3. Select or enter the login you wish to use and press `Enter`.

Reviewer 3: "enter the login" should maybe be "select or enter the `ssh` user you wish to login as

4. A new tab will open with a shell session on the chosen server.

Alternatively, you can type `tsh ssh` into the command bar and the autocompletion will guide you to

> connect to a server.

Reviewer 3: will guide you "through the server connection process" or "will guide you through connecting to a server" or "will help you connect to a server"

4. Select or enter the database user you with to use and press `Enter`.
5. A new tab will open with a new connection established between your device and the database
   server.

This connection will remain active until you click the Close Connection button or close Teleport
Connect. The port number will persist between app restarts – you can set up your favorite client
without worrying about the port suddenly changing.

### With a GUI Client

To connect with a GUI client, follow the instructions in the database connection tab under the
Connect with GUI section.

### With a CLI Client

The database connection tab shows the command that can be used to connect to the database. You can
modify the database name of the connection and then click the Run button to open a new
terminal tab with that command executed.

## Connecting to Multiple Clusters

After you add your first cluster, you can keep adding more by opening the profile selector in the
top right and clicking at the *+ Add another cluster* button.

At the moment Teleport Connect supports only one user per cluster. To log in as a different user,
log out of the cluster first.

## Restarting and Reconnecting

> Teleport Connect remembers the tabs that you had open and will ask you on restart if you want to

Reviewer 3: Not sure if this change is needed, but maybe reword it around "the previous session" or something around the word session since the modal that pops up when you open says "start new session" or reopen. lets use the "session" wording somewhere. "Connect session" maybe?

Author: Yeah that makes sense, also that sentence was super long. I changed it to the following:

> Before closing, Teleport Connect will remember the tabs that you had open at the end of the previous
> session. Next time you open it, Connect will ask you if you want to reopen those tabs. If you agree,
> Connect will restore connections to all resources that were active before you closed the app.

That "If you agree" part sounds weird to me though, now that I think about it.

reopen them. If you agree, Connect will restore connections to all resources that were active before
you closed the app.

At the moment terminal tabs are restored only as far as the working directory is concerned. Teleport
Connect doesn't attempt to re-execute commands that were in progress when the app was closed.

## Troubleshooting

In an attempt to resolve issues with a cluster you can try logging out of it, closing the app and
logging in again.

To force the app to log you out of all clusters, close the app and remove the `~/Library/Application
Support/Teleport Connect/tsh` folder. Removing the file under `~/Library/Application
Support/Teleport Connect/app_state.json` will clear all remembered tabs and connections.

### Submitting an Issue

To submit an issue, click the Submit Feedback button at the bottom right (the speech bubble symbol)

> and follow the Submit a Bug link.

Reviewer 3: lets highlight or quote "Submit a Bug"

## Connecting to an SSH server

1. Open a tab with cluster resources by clicking on the plus symbol at the right end of the tab bar.
   You can also press `Cmd + T` to achieve the same result.
2. Look for the SSH server you want to connect to and click the Connect button to the right.
3. Select or enter the SSH user you wish to log in as and press `Enter`.
4. A new tab will open with a shell session on the chosen server.

Alternatively, you can type `tsh ssh` into the command bar and the autocompletion will help you
connect to a server.

## Connecting to a database

1. Open a tab with cluster resources by clicking on the plus symbol at the end of the tab bar. You
   can also press `Cmd + T` to achieve the same result.
2. Select the Databases section.
3. Look for the database server you wish to connect to and click the Connect button to the right.
4. Select or enter the database user you with to use and press `Enter`.
5. A new tab will open with a new connection established between your device and the database
   server.

This connection will remain active until you click the Close Connection button or close Teleport

> Connect. The port number will persist between app restarts – you can set up your favorite client

Reviewer: ```suggestion
Connect. The port number will persist between app restarts—you can set up your favorite client
```
I think this looks better as an em dash (minor nit)

## Restarting and Reconnecting

Before closing, Teleport Connect will remember the tabs that you had open at the end of the previous
session. Next time you open it, Connect will ask you if you want to reopen those tabs. If you agree,
Connect will restore connections to all resources that were active before you closed the app.

> At the moment terminal tabs are restored only as far as the working directory is concerned. Teleport

Reviewer 3: ```suggestion
At the moment, terminal tabs are restored only as far as the working directory is concerned. Teleport
```
